doc/mkdssp.1
.TH mkdssp 1 "18-apr-2012" "version 2.0.4" "USER COMMANDS"
.SH NAME
mkdssp \- Calculate secondary structure for proteins in a PDB file
.SH SYNOPSIS
.B mkdssp
[OPTION] pdbfile [dsspfile]
.SH DESCRIPTION
The \fImkdssp\fP program was originally designed by Wolfgang Kabsch and
Chris Sander to standardize secondary structure assignment.
DSSP is a database of secondary structure assignments (and much more)
for all protein entries in the Protein Data Bank (PDB) and \fImkdssp\fP
is the application that calculates the DSSP entries from PDB entries.
Please note that \fImkdssp\fP \fBdoes not predict\fP secondary structure.
.SH OPTIONS
If you invoke \fImkdssp\fP with only one parameter, it will be
interpreted as the PDB file to process and output will be sent to
stdout. If a second parameter is specified this is interpreted as
the name of the DSSP file to create. Both the input and the output
file names may have either \.gz or \.bz2 as extension resulting in
the proper compression.
.TP
\fB\-i\fR, \fB\-\-input\fR filename
The file name of a
.I PDB
formatted file containing the protein structure data. This file
may be a file compressed by gzip or bzip2.
.TP
\fB\-o\fR, \fB\-\-output\fR filename
The file name of a
.I DSSP
file to create. If the filename ends in .gz or .bz2 a compressed
file is created.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Write out diagnositic information.
.TP
\fB\-\-version\fR
Print the version number and exit.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print the help message and exit.
The directory containing the parser scripts for mrs.
.SH THEORY
The DSSP program works by calculating the most likely secondary
structure assignment given the 3D structure of a protein. It does
this by reading the position of the atoms in a protein (the ATOM
records in a PDB file) followed by calculation of the H\-bond energy
between all atoms. The best two H\-bonds for each atom are then used
to determine the most likely class of secondary structure for each
residue in the protein.
.P
This means you do need to have a full and valid 3D structure for
a protein to be able to calculate the secondary structure.
There's no magic in DSSP, so e.g. it cannot guess the secondary
structure for a mutated protein for which you don't have the 3D
structure.
.SH DSSP FILE FORMAT
The header part of each DSSP file is self explaining, it contains
some of the information copied over from the PDB file and there
are some statistics gathered while calculating the secondary structure.
.P
The second half of the file contains the calculated secondary
structure information per residue. What follows is a brief
explanation for each column.
.TS
expand tab(;);
LfB LfB s.
Column Name;Description
_
.T&
L L s.
#;The residue number as counted by mkdssp
RESIDUE;T{
The residue number as specified by the PDB file followed by a chain identifier.
T}
AA;T{
The one letter code for the amino acid. If this letter is lower
case this means this is a cysteine that form a sulfur bridge
with the other amino acid in this column with the same lower
case letter.
T}
.T&
lt l s
c cfB lfB
c c l.
STRUCTURE;T{
This is a complex column containing multiple sub columns.
The first column contains a letter indicating the secondary
structure assigned to this residue. Valid values are:
T}
\^;Code;Description
\^;H;Alpha Helix
\^;B;Beta Bridge
\^;E;Strand
\^;G;Helix\-3
\^;I;Helix\-5
\^;T;Turn
\^;S;Bend
.T&
l l s.
\^;T{
What follows are three column indicating for each of the three
helix types (3, 4 and 5) whether this residue is a candidate in
forming this helix. A \fB>\fR character indicates it starts a helix,
a number indicates it is inside such a helix and a \fB<\fR character means
it ends the helix.
T}
\^;T{
The next column contains a S character if this residue is a possible bend.
T}
\^;T{
Then there's a column indicating the chirality and this can either be
positive or negative (i.e. the alpha torsion is either positive or negative).
T}
\^;T{
The last two columns contain beta bridge labels. Lower case here means
parallel bridge and thus upper case means anti parallel.
T}
.T&
L L s.
BP1 and BP2;T{
The first and second bridge pair candidate, this is followed by a letter
indicating the sheet.
T}
ACC;T{
The accessibility of this residue, this is the surface area expressed in
square Ångstrom that can be accessed by a water molecule.
T}
N\-H\-\->O..O\-\->H\-N;T{
Four columns, they give for each residue the H\-bond energy with another
residue where the current residue is either acceptor or donor. Each
column contains two numbers, the first is an offset from the current
residue to the partner residue in this H\-bond (in DSSP numbering), the
second number is the calculated energy for this H\-bond.
T}
TCO;T{
The cosine of the angle between C=O of the current residue and C=O of
previous residue. For alpha\-helices, TCO is near +1, for beta\-sheets
TCO is near \-1. Not used for structure definition.
T}
Kappa;T{
The virtual bond angle (bend angle) defined by the three C\-alpha atoms
of the residues current \- 2, current and current + 2. Used to define
bend (structure code 'S').
T}
PHI and PSI;T{
IUPAC peptide backbone torsion angles.
T}
X\-CA, Y\-CA and Z\-CA;T{
The C\-alpha coordinates
T}
.TE

.SH HISTORY
The original DSSP application was written by Wolfgang Kabsch and
Chris Sander in Pascal. This version is a complete rewrite in C++
based on the original source code. A few bugs have been fixed since
and the algorithms have been tweaked here and there.
.SH TODO
The code desperately needs an update. The first thing that needs
implementing is the improved recognition of pi\-helices. A second
improvement would be to use angle dependent H\-bond energy calculation.
.SH BUGS
If you find any, please let me know.
.SH AUTHOR
Maarten L. Hekkelman (m.hekkelman (at) cmbi.ru.nl)


src/mkdssp.cpp
// Copyright Maarten L. Hekkelman, Radboud University 2008-2011.
//   Distributed under the Boost Software License, Version 1.0.
//       (See accompanying file LICENSE_1_0.txt or copy at    
//             http://www.boost.org/LICENSE_1_0.txt)      
//
// A DSSP reimplementation

#include "mas.h"

#if defined(_MSC_VER)
#include <conio.h>
#include <ctype.h>
#endif

#include <fstream>

#include <boost/program_options.hpp>
#include <boost/iostreams/filtering_stream.hpp>
#include <boost/iostreams/copy.hpp>
#if defined USE_COMPRESSION
#include <boost/iostreams/filter/bzip2.hpp>
#include <boost/iostreams/filter/gzip.hpp>
#endif
#include <boost/algorithm/string.hpp>

#include "dssp.h"
#include "structure.h"

using namespace std;
namespace po = boost::program_options;
namespace io = boost::iostreams;
namespace ba = boost::algorithm;

int MULTI_THREADED = 1, VERBOSE;

int main(int argc, char* argv[])
{
	try
	{
		po::options_description desc("mkdssp " VERSION " options");
		desc.add_options()
			("help,h",							 "Display help message")
			("input,i",		po::value<string>(), "Input file")
			("output,o",	po::value<string>(), "Output file, use 'stdout' to output to screen")
			("verbose,v",						 "Verbose output")
			("version",							 "Print version")
			("debug,d",		po::value<int>(),	 "Debug level (for even more verbose output)")
			;
	
		po::positional_options_description p;
		p.add("input", 1);
		p.add("output", 2);
	
		po::variables_map vm;
		po::store(po::command_line_parser(argc, argv).options(desc).positional(p).run(), vm);
		po::notify(vm);

		if (vm.count("version"))
		{
			cout << "mkdssp version " VERSION << endl;
			exit(0);
		}

		if (vm.count("help") or not vm.count("input"))
		{
			cerr << desc << endl
				 << endl
				 << "Examples: " << endl
				 << endl
				 << "To calculate the secondary structure for the file 1crn.pdb and" << endl
				 << "write the result to a file called 1crn.dssp, you type:" << endl
				 << endl
				 << "  " << argv[0] << "-i 1crn.pdb -o 1crn.dssp" << endl
				 << endl;
#if defined(_MSC_VER)
			cerr << endl
				 << "MKDSSP is a command line application, use the 'Command prompt' application" << endl
				 << "to start " << argv[0] << " You can find the 'Command prompt' in the Start menu:" << endl
				 << endl
				 << "Start => Accessories => Command prompt" << endl
				 << endl
				 << endl
				 << "Press any key to continue..." << endl;
			char ch = _getch();
#endif
			exit(1);
		}

		VERBOSE = vm.count("verbose") != 0;
		if (vm.count("debug"))
			VERBOSE = vm["debug"].as<int>();
		
		string input = vm["input"].as<string>();

		ifstream infile(input.c_str(), ios_base::in | ios_base::binary);
		if (not infile.is_open())
			throw runtime_error("No such file");
		
		io::filtering_stream<io::input> in;
		
#if defined USE_COMPRESSION
		if (ba::ends_with(input, ".bz2"))
			in.push(io::bzip2_decompressor());
		else if (ba::ends_with(input, ".gz"))
			in.push(io::gzip_decompressor());
#endif
		
		in.push(infile);
	
		// OK, we've got the file, now create a protein
		MProtein a(in);
		
		// then calculate the secondary structure
		a.CalculateSecondaryStructure();
	
		// and finally report the secondary structure in the DSSP format
		// either to cout or an (optionally compressed) file.
		if (vm.count("output"))
		{
			string output = vm["output"].as<string>();
			
			ofstream outfile(output.c_str(), ios_base::out|ios_base::trunc|ios_base::binary);
			if (not outfile.is_open())
				throw runtime_error("could not create output file");
			
			io::filtering_stream<io::output> out;
#if defined USE_COMPRESSION
			if (ba::ends_with(output, ".bz2"))
				out.push(io::bzip2_compressor());
			else if (ba::ends_with(output, ".gz"))
				out.push(io::gzip_compressor());
#endif
			out.push(outfile);
			
			WriteDSSP(a, out);
		}
		else
			WriteDSSP(a, cout);
	}
	catch (const exception& e)
	{
		cerr << "DSSP could not be created due to an error:" << endl
			 << e.what() << endl;
		exit(1);
	}
	
	return 0;
}

